File: scripts/update_archive.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;

// --- KONFIGURATION ---
final String githubOrg = Platform.environment['GITHUB_ORG'] ?? '';
final String token = Platform.environment['GH_TOKEN'] ?? '';
final Map<String, String> headers = {
  'Authorization': 'token $token',
  'Accept': 'application/vnd.github.v3+json',
};

const String readmePath = 'profile/README.md';

class Course {
  final String title;
  final String longTitle;
  final String type;
  final String year;
  final String grade;
  final String campNumber;
  final String url;

  Course({
    required this.title,
    required this.longTitle,
    required this.type,
    required this.year,
    required this.grade,
    required this.campNumber,
    required this.url,
  });

  // Getter til at bestemme kategorien
  String get category =>
      type.toLowerCase().contains('aspirant') ? 'Aspirant' : 'Folkeskole';

  // Getter til Folkeskole visning
  String get folkeskoleDisplay => "$grade. klasse, camp $campNumber";

  factory Course.fromJson(Map<String, dynamic> json, String repoUrl) {
    return Course(
      title: json['display_name'] ?? 'Navn mangler',
      longTitle:
          json['long_display_name'] ?? json['display_name'] ?? 'Navn mangler',
      type: json['type'] ?? 'Andet',
      year: json['year']?.toString() ?? 'Ukendt år',
      grade: json['grade']?.toString() ?? '?',
      campNumber: json['camp_number']?.toString() ?? '?',
      url: repoUrl,
    );
  }
}

Future<List<Course>> getCourseData() async {
  List<Course> allCourses = [];
  int page = 1;

  while (true) {
    final reposUrl =
        'https://api.github.com/orgs/$githubOrg/repos?per_page=100&page=$page';
    final response = await http.get(Uri.parse(reposUrl), headers: headers);

    if (response.statusCode != 200) break;

    final List repos = jsonDecode(response.body);
    if (repos.isEmpty) break;

    for (var repo in repos) {
      final metaUrl =
          'https://api.github.com/repos/$githubOrg/${repo['name']}/contents/meta.json';
      final metaRes = await http.get(Uri.parse(metaUrl), headers: headers);

      if (metaRes.statusCode == 200) {
        try {
          final metaJson = jsonDecode(metaRes.body);
          // GitHub base64 koder indholdet, vi skal dekode det
          final content = utf8.decode(
            base64.decode(metaJson['content'].replaceAll('\n', '')),
          );
          final Map<String, dynamic> metaData = jsonDecode(content);

          allCourses.add(Course.fromJson(metaData, repo['html_url']));
        } catch (e) {
          print('Fejl ved parsing af meta.json i ${repo['name']}: $e');
        }
      }
    }
    page++;
  }
  return allCourses;
}

String buildMarkdown(List<Course> courses) {
  // Struktur: Type -> År -> Liste af kurser
  Map<String, Map<String, List<Course>>> tree = {};

  for (var c in courses) {
    tree.putIfAbsent(c.category, () => {});
    tree[c.category]!.putIfAbsent(c.year, () => []);
    tree[c.category]![c.year]!.add(c);
  }

  var sb = StringBuffer("# Kursus Arkiv\n\n");

  // Sorter Typer
  var sortedTypes = tree.keys.toList()..sort();
  for (var type in sortedTypes) {
    sb.writeln("## $type\n");

    // Sorter År (nyeste først)
    var sortedYears = tree[type]!.keys.toList()..sort((a, b) => b.compareTo(a));
    for (var year in sortedYears) {
      sb.writeln("<details>\n  <summary><h3>Årstal: $year</h3></summary>\n");

      var coursesInYear = tree[type]![year]!;

      if (type == "Aspirant") {
        // Grupper efter Samling
        Map<String, List<Course>> camps = {};
        for (var c in coursesInYear) {
          String campName = "Samling ${c.campNumber}";
          camps.putIfAbsent(campName, () => []);
          camps[campName]!.add(c);
        }

        var sortedCamps = camps.keys.toList()..sort();
        for (var camp in sortedCamps) {
          sb.writeln("  <details style='margin-left: 20px;'>");
          sb.writeln("    <summary><h4>$camp</h4></summary>\n");
          for (var c in camps[camp]!) {
            sb.writeln("    * [${c.title}](${c.url})");
          }
          sb.writeln("  </details>");
        }
      } else {
        // Folkeskole: Sorter efter longTitle
        coursesInYear.sort((a, b) => a.longTitle.compareTo(b.longTitle));
        for (var c in coursesInYear) {
          sb.writeln(
            "  * [${c.longTitle}](${c.url}) — *${c.folkeskoleDisplay}*",
          );
        }
      }
      sb.writeln("</details>\n");
    }
    sb.writeln("---\n");
  }
  return sb.toString();
}

void main() async {
  if (githubOrg.isEmpty || token.isEmpty) {
    print('Fejl: GITHUB_ORG eller GH_TOKEN mangler i environment.');
    exit(1);
  }

  print('Henter data fra GitHub...');
  final data = await getCourseData();

  print('Bygger Markdown...');
  final markdown = buildMarkdown(data);

  print('Gemmer til $readmePath...');
  final file = File(readmePath);
  await file.parent.create(recursive: true);
  await file.writeAsString(markdown);

  print('Succes!');
}

File: update_archive.py
import os
import requests
import base64
import json

# --- KONFIGURATION ---
GITHUB_ORG = os.getenv("GITHUB_ORG")
TOKEN = os.getenv("GH_TOKEN")
HEADERS = {"Authorization": f"token {TOKEN}"}

README_PATH = "profile/README.md"


def get_course_data():
    """
    Henter kursusdata fra alle repositories i GitHub-organisationen.
     - For hver repository, tjekker den for en 'meta.json' fil.
     - Hvis filen findes, dekoder den indholdet og tilføjer det til kursuslisten.
     - Returnerer en liste af kursusmetadata.
    """
    all_courses = []
    page = 1
    while True:
        repos_url = (
            f"https://api.github.com/orgs/{GITHUB_ORG}/repos?per_page=100&page={page}"
        )
        res = requests.get(repos_url, headers=HEADERS, timeout=30)
        repos = res.json()

        if not repos or len(repos) == 0:
            break

        for repo in repos:
            meta_url = f"https://api.github.com/repos/{GITHUB_ORG}/{repo['name']}/contents/meta.json"
            meta_res = requests.get(meta_url, headers=HEADERS, timeout=30)

            if meta_res.status_code == 200:
                content = base64.b64decode(meta_res.json()["content"]).decode("utf-8")
                meta = json.loads(content)
                meta["url"] = repo["html_url"]
                all_courses.append(meta)
        page += 1
    return all_courses


# TODO: sørg for at det ikke lukket helt hvis en meta.json er forkert formateret,
# TODO: sørg for at der til hvert kursus er den rigtige information, lige nu er informationen i loopet, hvilket er forker
def build_markdown(courses):
    """
    Bygger en Markdown-struktur baseret på kursusdataene.
     - Kurserne grupperes først efter type (Aspirant eller Folkeskole).
     - Derefter grupperes de efter årstal.
     - For Aspirant-kurser, grupperes de yderligere efter samling.
     - For Folkeskole-kurser, vises de direkte under årstallet.
     - Returnerer den færdige Markdown-struktur som en streng.
    """
    # Vi bygger et træ: Type -> Årstal -> (Evt. Samling) -> Kurser
    tree = {}
    for c in courses:
        # Standardisering af data
        raw_type = str(c.get("type", "Andet")).lower()
        c_type = "Aspirant" if "aspirant" in raw_type else "Folkeskole"
        grade = str(c.get("grade", ""))
        year = c.get("year", "Ukendt år")
        camp_number = str(c.get("camp_number", "?"))

        if c_type not in tree:
            tree[c_type] = {}
        if year not in tree[c_type]:
            tree[c_type][
                year
            ] = []  # Vi starter med en liste, som vi kan strukturere senere

        tree[c_type][year].append(c)

    md = "# Kursus Arkiv\n\n"

    # Sorter typer (Aspirant øverst, så Folkeskole)
    for c_type in sorted(tree.keys()):
        md += f"## {c_type}\n\n"

        # Sorter årstal (nyeste først)
        for year in sorted(tree[c_type].keys(), reverse=True):
            md += f"<details>\n  <summary><h3>Årstal: {year}</h3></summary>\n\n"

            courses_in_year = tree[c_type][year]

            if c_type == "Aspirant":
                # LOGIK FOR ASPIRANT: Grupper efter Samling
                camps = {}
                for c in courses_in_year:
                    camp_name = f"Samling {camp_number}"
                    if camp_name not in camps:
                        camps[camp_name] = []
                    camps[camp_name].append(c)

                for camp_name in sorted(camps.keys()):
                    md += f"  <details style='margin-left: 20px;'>\n    <summary><h4>{camp_name}</h4></summary>\n\n"
                    for repo in camps[camp_name]:
                        name = repo.get(
                            "display_name", repo.get("name", "Navn mangler")
                        )
                        md += f"  * [{name}]({repo['url']})\n"
                    md += "  </details>\n"

            else:
                folkeskole_display = f"{grade}. klasse, camp {camp_number} *()*"
                # LOGIK FOR FOLKESKOLE: Ingen samlinger, brug 'long_display_name'
                # Vi sorterer dem alfabetisk efter det lange navn
                sorted_folkeskole = sorted(
                    courses_in_year, key=lambda x: x.get("long_display_name", "")
                )
                for repo in sorted_folkeskole:
                    md += f"  * [{folkeskole_display}]({repo['url']})\n"

            md += "</details>\n\n"

        md += "---\n\n"
    return md


def overwrite_readme(new_content):
    """
    Overskriver README.md med det nye indhold.
    Hvis filen ikke findes, oprettes den.
    """
    os.makedirs(os.path.dirname(README_PATH), exist_ok=True)
    with open(README_PATH, "w", encoding="utf-8") as f:
        f.write(new_content)


if __name__ == "__main__":
    data = get_course_data()
    archive_md = build_markdown(data)
    overwrite_readme(archive_md)
    print(f"Succes! {README_PATH} er opdateret.")
